src/Uno.Core/Collections/ImmutableList.cs

```diff
@@ -24,33 +26,203 @@ namespace Uno.Collections
 		/// Provides an empty list
 		/// </summary>
-		public static ImmutableList<T> Empty {
-			get { return _empty; }
-		}
+		public static ImmutableList<T> Empty => _empty;
 
 		/// <summary>
 		/// Initializes the list with the provided array.
 		/// </summary>
 		/// <param name="data"></param>
 		public ImmutableList(T[] data)
 		{
-			this.data = data;
+			_data = new T[data.Length];
+			Array.Copy(data, _data, data.Length);
+		}
+
+		public IImmutableList<T> Clear()
+		{
+			return Empty;
+		}
+
+		public int IndexOf(T item, int index, int count, IEqualityComparer<T> equalityComparer)
+		{
+			var comparer = equalityComparer ?? EqualityComparer<T>.Default;
+			for (var i = index; i < index + count; i++)
+			{
+				if (comparer.Equals(_data[i], item))
+				{
+					return i;
+				}
+			}
+			return -1;
+		}
+
+		public int LastIndexOf(T item, int index, int count, IEqualityComparer<T> equalityComparer)
+		{
+			var comparer = equalityComparer ?? EqualityComparer<T>.Default;
+			for (var i = index + count - 1; i >= index; i--)
+			{
+				if (comparer.Equals(_data[i], item))
+				{
+					return i;
+				}
+			}
+			return -1;
+		}
+
+		IImmutableList<T> IImmutableList<T>.Add(T value)
+		{
+			return Add(value);
+		}
+
+		public IImmutableList<T> AddRange(IEnumerable<T> items)
+		{
+			var itemsToAdd = items.ToArray();
+			var newData = new T[_data.Length + itemsToAdd.Length];
+			Array.Copy(_data, newData, _data.Length);
+			Array.Copy(itemsToAdd, 0, newData, _data.Length, itemsToAdd.Length);
+			return new ImmutableList<T>(newData);
+		}
+
+		public IImmutableList<T> Insert(int index, T element)
+		{
+			var newData = new T[_data.Length + 1];
+			newData[index] = element; // will throw out-of-range exception if index is not valid
+
+			if (index > 0)
+			{
+				Array.Copy(_data, newData, index);
+			}
+			if (index + 1 < _data.Length)
+			{
+				Array.Copy(_data, index, newData, index + 1, _data.Length - index);
+			}
+			return new ImmutableList<T>(newData);
+		}
+
+		public IImmutableList<T> InsertRange(int index, IEnumerable<T> items)
+		{
+			var insertedItems = items.ToArray();
+			if (insertedItems.Length == 0)
+			{
+				return this; // nothing to insert
+			}
+
+			var newData = new T[_data.Length + insertedItems.Length];
+
+			if (index > 0)
+			{
+				Array.Copy(_data, newData, index);
+			}
+
+			Array.Copy(insertedItems, 0, newData, index, insertedItems.Length);
+
+			if (index + 1 < _data.Length)
+			{
+				Array.Copy(_data, index, newData, index + insertedItems.Length, _data.Length - index);
+			}
+			return new ImmutableList<T>(newData);
+		}
+
+		public IImmutableList<T> Remove(T value, IEqualityComparer<T> equalityComparer)
+		{
+			var comparer = equalityComparer ?? EqualityComparer<T>.Default;
+			return RemoveAll(x => comparer.Equals(x, value));
+
+		}
+
+		public IImmutableList<T> RemoveAll(Predicate<T> match)
+		{
+			var newData = _data.Where(x => !match(x)).ToArray();
+			return new ImmutableList<T>(newData);
```

Comment:
If there is not matches, we should return this

Reply:
I'll do it.